### ISSUE TITLE
Problem with v8 model fine tuning/custom training

### ISSUE THREAD

**Issue by veeyuen** [ultralytics/ultralytics#1686]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I attempted to fine tune the existing standard nano model with a custom dataset but the previous classes trained on the COCO dataset have been completely over-written and no longer recognised.

I tried to seach for a way to freeze the backbone in order to retain most of the existing weights, but it appears that there is no ability to do so under v8, unlike in v5.

Am I missing something when it comes to the setup during training?  If anyone has any suggestions it would be greatly appreciated

**Comment by glenn-jocher** [comment]
In YOLOv8, you can use the `--weights` argument to load the pre-trained COCO weights and then use `--cfg <cfgfile> --weights <weightsfile> --img 640 --batch 16 --epochs 300 --data coco.yaml --nosave --cache` to fine-tune on your custom dataset by specifying your custom data in a `.yaml` file. 

To freeze the backbone weights during training, you can set `stopbackward=True` during `train()` function call. This will prevent the gradients from propagating back through the backbone layers, which effectively freezes their weights. 

Here is an example of using `stopbackward=True` during `train()` function call:

```python
from models.experimental import attempt_load
from utils.general import check_dataset

weights = 'path/to/coco/weights.pt'
cfg = 'path/to/yolov8.cfg'
data = 'path/to/custom.yaml'
batch_size = 16
epochs = 300
img_size = 640

# Load model
model = attempt_load(weights, map_location=device)  # load FP32 model
stride = int(model.stride.max())  # model stride

# Freeze backbone
model.model[-1].stopbackward = True

# Check that the dataset is valid
check_dataset(data)

# Train the model
model.train(data,
            epochs=epochs,
            batch_size=batch_size,
            imgsz=img_size,
            stride=stride)
``

**Comment by glenn-jocher** [comment]
It seems like you are experiencing an issue with catastrophic forgetting, which is a common challenge when fine-tuning deep learning models on new datasets. Since you mentioned using the v8 version, I assume you are working with YOLOv8 for object detection.

In YOLOv5, you can freeze layers of the backbone to prevent weights from being overwritten, which helps retain knowledge from the COCO dataset. Unfortunately, YOLOv8 does not have a straightforward option to freeze layers during training. However, you can still take some steps to mitigate this issue:

Lower learning rate: Reduce the learning rate to a smaller value, so the weights don't update too aggressively during fine-tuning. This may help retain the knowledge from the COCO dataset.

Gradual unfreezing: You can implement this yourself by manually freezing and unfreezing layers during training. Start by freezing all the layers except the last few, then gradually unfreeze more layers as training progresses. This technique helps retain the earlier knowledge while still allowing the model to learn new information.

Mix datasets: Combine your custom dataset with a subset of the COCO dataset during fine-tuning. By doing this, you encourage the model to continue learning about the COCO dataset classes while also learning about the new classes in your custom dataset. This approach can help reduce catastrophic forgetting.

Knowledge distillation: Train a student model on your custom dataset using the logits from the pre-trained model as soft targets. This approach can help the new model retain knowledge from the pre-trained model while still learning new information from the custom dataset.

Remember that the YOLO family of models are continuously evolving, and new versions or forks may have different capabilities. Keep an eye on the official repository and related forks for updates that may include the ability to freeze layers or other techniques to address catastrophic forgetting.

**Comment by veeyuen** [comment]
Hi Glenn,

Thanks for your reply.  Am I able to identify specific layers to freeze/unfreeze in YOLOv8 via the CLI or do I need to customize the code in order to do that?  I didn't expect to encounter this issue of catastrophic forgetting in v8 to such an extent.

**Comment by glenn-jocher** [comment]
Hello @veeyuen, unfortunately, YOLOv8 doesn't have a straightforward option to freeze layers during training via the CLI. As I mentioned earlier, you would need to manually modify the code to implement this feature. 

However, you can lower the learning rate during training to reduce the impact of catastrophic forgetting, as well as try the other techniques that I mentioned earlier, which includes mixing datasets and gradual unfreezing.

YOLOv8 has several other features, such as an improved anchor-free architecture, better performance, and new performance benchmarks. With that being said, catastrophic forgetting is a common challenge when fine-tuning deep learning models on new datasets. It is important to be prepared to handle it when fine-tuning any pre-trained models.

Please let me know if you have any further questions or concerns.

**Comment by veeyuen** [comment]
@glenn-jocher , thanks will try all your suggestions, and see which combinations work best against catastrophic forgetting

**Comment by glenn-jocher** [comment]
You're welcome, @veeyuen! I hope these suggestions help manage catastrophic forgetting in your fine-tuning process. Don't hesitate to ask if you have any other questions or concerns. Good luck with your experimentation!

**Comment by veeyuen** [comment]
Hi @glenn-jocher,  has anyone looked into implementing Elastic Weight Consolidation in YOLO?  Do you think that woud be beneficial?

**Comment by glenn-jocher** [comment]
@veeyuen elastic Weight Consolidation (EWC) is a popular technique used for mitigating catastrophic forgetting in deep learning models. The premise of EWC is to place strong priors on the model parameters that correspond to the previously learned knowledge (in this case, knowledge from the pre-trained COCO dataset), discouraging the network from updating these weights too much during fine-tuning on new data.

While I am unaware of any current efforts to implement EWC explicitly in the YOLOv8 model, it is a promising technique that has shown effectiveness in various other deep learning models. So it could be a worthwhile area of investigation for improving the performance of YOLOv8 and reducing catastrophic forgetting.

Just note that depending on the nature of the data, different techniques such as distillation or early stopping may yield better performance, so it's important to experiment and see what works best for your specific use case.

**Comment by veeyuen** [comment]
@glenn-jocher After studying the code in detail I just figured out how to freeze each layer, hence I am now running a few empirical tests with different hyperparameters and active layers, akin to a grid search.  

Concurrently, I am also looking into more theoretical foundations for dealing with catastrophic forgeting, hence EWC, in the event those efforts fall short.  In your opinion would it be difficult to enhance the current code to take into account EWC?

Thanks for your advice.

**Comment by glenn-jocher** [comment]
@veeyuen that's great to hear that you were able to figure out how to freeze individual layers in the YOLOv8 codebase! And it's also a good idea to look into more theoretical foundations for dealing with catastrophic forgetting, such as EWC.

In theory, it is not too difficult to implement EWC in the YOLOv8 codebase. The main idea is to estimate a Fisher Information Matrix (FIM) for each learned parameter during the pre-training process. The FIM is then used to compute a regularization term that encourages the model to preserve the original knowledge during fine-tuning on the new dataset. 

The key challenge in implementing EWC is estimating the FIM for the pre-trained model. This involves computing the second-derivative of the loss function with respect to each learned parameter, which can be computationally expensive. Additionally, computing the FIM for a large model such as YOLOv8 may require a significant amount of memory. However, there are existing frameworks such as PyTorch that provide automatic differentiation tools to compute the second derivatives efficiently, which can help streamline the implementation of EWC.

Overall, while implementing EWC in YOLOv8 may require some computational resources, it is a promising technique for reducing catastrophic forgetting and improving performance during fine-tuning.
